lzy/execution-env/src/main/java/ai/lzy/env/base/DockerEnvironment.java

```diff
@@ -243,11 +248,21 @@ public InputStream err() {
             }
 
             @Override
-            public int waitFor() throws InterruptedException {
+            public int waitFor() throws InterruptedException, OomKilledException {
                 try {
                     feature.get();
-                    return Math.toIntExact(retry.executeSupplier(() -> client.inspectExecCmd(exec.getId()).exec())
+                    var rc = Math.toIntExact(retry.executeSupplier(() -> client.inspectExecCmd(exec.getId()).exec())
                         .getExitCodeLong());
+
+                    if (rc == 0) {
+                        return 0;
+                    }
+
+                    if (isOomKilled()) {
```

Comment:
1. you should check OOM before retcode
2. right now you do `inspect` twice: on line 254 and line 385